mantidproject/mantid
# API function to replace workspace being shown by an instrument window

**Description of work.**
This PR adds an API function to replace the workspace that is being shown by an instrument window.
When called, the view is reset, the color and integration range are updated and the window's title is changed to new values. Only the instrument render mode is kept the same.
This is a step in the development of a future widget, and as such is not really useful on its own.

**To test:**
The instrument widget not being compliant with the MVP model, this function unfortunately needs manual testing.
The following strange syntax regarding threads is due to issues with the way they communicate in Workbench's embedded python interpreter.

To test: 
- Add `ExternalData/Testing/Data/UnitTest/ILL/D33/` to the list of user directories.
- Run the following script. During the pauses provided by the sleeps, manipulate the instrument view and check that everything - integration range, color bar, view direction, pick tab - is reset when the view changes.
```
from mantid.simpleapi import CreateSampleWorkspace
from mantidqt.widgets.instrumentview.presenter import InstrumentViewPresenter
from mantidqt.utils.qt.qappthreadcall import QAppThreadCall

from time import sleep

def show_and_replace_instrument_view(first_ws, first_name, second_name, new_window_name=None):
    # For thread related reasons, the instrument view and the replace_ws calls need to be made on the same function,
    # using QAppthreadCall.
    # This limitation is only present for scripts run in the workbench's embedded script window, and should not be of concern
    # in the final use case of this function.
    
    iv = QAppThreadCall(InstrumentViewPresenter)(first_ws)
    QAppThreadCall(iv.show_view)()
    
    while True:
        sleep(8)
        QAppThreadCall(iv.replace_workspace)(second_name, new_window_name)
        sleep(2)
        QAppThreadCall(iv.replace_workspace)(first_name)
    
first = Load(Filename="093410", OutputWorkspace="093410")
second = LoadEmptyInstrument(InstrumentName="d16", OutputWorkspace="ws")

show_and_replace_instrument_view(first, "093410", "ws", "Empty_D16")
```

Fixes #29846 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
